### ISSUE TITLE
T1 and extended length command failure

### ISSUE THREAD

**Issue by ijsf** [wookey-project/libiso7816#13]
Hi, I've been using `libiso7816` with my smartcard application at T=1 configuration. Short commands (<255 byte length) seem to work well, but as soon as an extended length command is sent, errors are shown:

```
SC_fsm_early_init(SC_MAP_AUTO);
SC_fsm_init(&card, 0 /* do_negotiate_pts */, 0 /* do_change_baudrate */, 2 /* do_force_protocol T=1 */, 0 /* do_force_etu */);

apdu.cla = 0x00; apdu.ins = requestType; apdu.p1 = apdu.p2 = 0x00;
apdu.le = 0xFF; apdu.send_le = 1;
SC_send_APDU(&apdu, &res, &card);
```

This is using the default init (though with T=1 forced as it is a T=1 card), and with a smartcard applet that is known to work with extended length commands in other programs.

Returns the following error:

```
[Smartcard T=1] Unexpected case: received other block than expected IBLOCK, or bad sequence number ...
```
and seems to fail here: https://github.com/wookey-project/libiso7816/blob/master/smartcard_iso7816.c#L1952

I can see that a RBLOCK is received at that point while the last block (out of 2 blocks) is being sent and therefore an IBLOCK is expected.

Not sure whats going on, but this seems to be a bug in the handling logic?

**Comment by ijsf** [comment]
It appears that there are actually several minimum delays (waits) missing in this code, one of which should be placed in this loop, possibly around here: https://github.com/wookey-project/libiso7816/blob/master/smartcard_iso7816.c#L1890

This is so the block guard time isn't violated when there are multiple blocks to send, e.g.:

```
SC_delay_etu(BGT_block_guard_time); /* Wait for the standardized Block Guard Time (22 ETU by default) */
```

Adding a wait seems to solve the problem of duplicate RBLOCKs being received. This still doesn't fix the issue 100% (applet reports that the length is wrong) but it solves one of the problems.

**Comment by ijsf** [comment]
Apparently the case for `apdu.send_le = 1` is not handled well when an extended command is used either in `libsmartcard` or `libiso7816` and causes `0x6700` (wrong length) to be returned. Changing this to `apdu.send_le = 0` or `apdu.send_le = 2` (extended response) will fix the issue along with the above fix, e.g.:

```
apdu.cla = 0x00; apdu.ins = requestType; apdu.p1 = apdu.p2 = 0x00;
apdu.le = 0xFF; apdu.send_le = 2;
SC_send_APDU(&apdu, &res, &card);
```

Hopefully this will help anyone out there just in case there are no maintainers active on this project anymore.

**Comment by rben-dev** [comment]
Hi @ijsf,

Sorry for answering after you found this by yourself, but yes: **extended APDUs** are specifically handled with `send_le=2`: this is actually somehow explained in the `libsmartcard` layer https://github.com/wookey-project/libsmartcard/blob/02f36344761d0619f98e88fb07c8a47cc6505f78/smartcard.c#L30 (the layer handling APDUs, the `libiso7816` handles the lower T0 and T1 TPDU layers).

Do you confirm that the issue is not fully solved only with `send_le=2`? Also, what kind of smartcard do you use in T1 mode? The library has been tested with short and extended APDUs on various T=0 and T=1 smartcards from various vendors, I am curious to see when this fails.

Also, regarding the wait delays, the guard time corresponding to the line  https://github.com/wookey-project/libiso7816/blob/master/smartcard_iso7816.c#L1890 should be here: https://github.com/wookey-project/libiso7816/blob/95a644a2108b06f2b4b162f58caac0fed2cc5a0a/smartcard_iso7816.c#L1984 (as it is the code executed after going out of the IBLOCKs sending loop).

Regards,

**Comment by ijsf** [comment]
Hi @rben-dev thanks for responding, I assumed the project was inactive but happy to see some response.

So the `send_le=2` change makes sense after checking the docs. The previous value I used is simply an invalid APDU case, though an error could help if a developer makes this mistake like me.

The actual issue is still there even if `send_le=2` is set.

This is because the for-loop doesn't have the mandatory Block Guard Time wait in case it loops over (for loop: https://github.com/wookey-project/libiso7816/blob/master/smartcard_iso7816.c#L1827). Note that the delay you're referring to is in a loop below that is only executed after all IBLOCKS have been sent. The BGT seems necessary between the IBLOCKS sends in the earlier for loop I'm referring to.

P.S. I'm using NXP J2E081.

**Comment by rben-dev** [comment]
Agreed that when looping on IBLOCKs there, the BGT must be respected in this `if` condition and it is absent.
Weird though that this fails / appears on a NXP JCOP as many of these cards have been tested with T=1 extended APDUs (maybe not the exact same reference though). When fixing this BGT, do you see other issues?

Would you accept to share the piece of applet handling the APDU that provokes the issue (not specifically all the applet, but a minimal working example): this will help to reproduce this :-) Also, if you mind sharing: on what platform is the reader implemented?

Regards,

**Comment by ijsf** [comment]
No other issues, it seems to be working fine now after adding the delay.

Unfortunately I can't really share parts of the applet, however you should be able to reproduce this with the Hello World sample in this repository: https://github.com/ph4r05/javacard-gradle-template.

The platform is a Teensy 4.0 (NXP iMXRT1062). I can imagine that the timing differences on the platforms may account for why you haven't seen this before, e.g.the loop on this platform is executed much faster.

**Comment by rben-dev** [comment]
Thanks for the feedback! Yes the idea behind my question was exactly what you mention: the timing difference between the platforms (we were testing on a STM32F4xx @168 MHz, clearly slower than your Teensy 4.0 @600 MHz) might explain why this specific missing BGT was "hidden".

Also, thanks for the applet reference: I will take a look and perform tests (ideally with another reader platform if the issue indeed lies in the timing difference ...).

Anyways, this should be fixed in commit 9e214497e8fe038f01f0dae61b13596397870b62: can you please check that everything is OK, and if so close the issue?

Regards,

**Comment by ijsf** [comment]
Thanks again. Looks good to me!
